perf/evp_fetch.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+#include <openssl/core_names.h>
+#include "perflib/perflib.h"
+
+#define NUM_CALLS_PER_TEST         100000
+
+OSSL_TIME *times;
+
+int err = 0;
+
+static int threadcount;
+
+static OSSL_LIB_CTX *ctx = NULL;
+
+#define ARRAY_SIZE(a)                               \
+  ((sizeof(a) / sizeof(*(a))))
+
+typedef enum {
+    FETCH_MD = 0,
+    FETCH_CIPHER,
+    FETCH_KDF,
+    FETCH_MAC,
+    FETCH_RAND
+} fetch_type_t;
+
+struct fetch_data_entry {
+    fetch_type_t ftype;
+    const char *alg;
+    const char *propq;
+};
+
+static struct fetch_data_entry fetch_entries[] = {
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA2_224, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA2_256, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA3_224, NULL},
+    {FETCH_MD, OSSL_DIGEST_NAME_SHA3_256, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_128_GCM_SIV, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_192_GCM_SIV, NULL},
+    {FETCH_CIPHER, OSSL_CIPHER_NAME_AES_256_GCM_SIV, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_HKDF, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_SCRYPT, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_KRB5KDF, NULL},
+    {FETCH_KDF, OSSL_KDF_NAME_KBKDF, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_BLAKE2BMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_CMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_GMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_HMAC, NULL},
+    {FETCH_MAC, OSSL_MAC_NAME_POLY1305, NULL},
+    {FETCH_RAND, "CTR-DRBG", NULL}
```

Comment:
More for the future: would storing timing information for each be useful?  I doubt there is a significant difference between each but it would be nice to know for sure.

Reply:
That might be useful.  Perhaps a cmd line option or argument to direct the test to only do fetches of a specific type (or even specific alg), so that we could test something particular, or generally get a sampling of all of them.

Reply:
updated to include this